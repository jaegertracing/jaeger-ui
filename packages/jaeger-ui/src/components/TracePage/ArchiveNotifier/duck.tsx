// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { createActions, handleActions, ActionFunctionAny } from 'redux-actions';

import { archiveTrace } from '../../../actions/jaeger-api';
import { ApiError } from '../../../types/api-error';
import { TracesArchive } from '../../../types/archive';
import generateActionTypes from '../../../utils/generate-action-types';

type ArchiveAction = {
  meta: {
    id: string;
  };
  payload?: ApiError | string;
};

const initialState: TracesArchive = {};

const actionTypes = generateActionTypes('@jaeger-ui/archive-trace', ['ACKNOWLEDGE']);

const fullActions = createActions({
  [actionTypes.ACKNOWLEDGE]: traceID => traceID,
});

export const actions: { [actionType: string]: ActionFunctionAny<any> } = {
  ...(fullActions as any).jaegerUi.archiveTrace,
  archiveTrace,
};

function acknowledge(state: TracesArchive, { payload }: ArchiveAction) {
  const traceID = typeof payload === 'string' ? payload : null;
  if (!traceID) {
    // make flow happy
    throw new Error('Invalid state, missing traceID for archive acknowledge');
  }
  const traceArchive = state[traceID];
  if (traceArchive && traceArchive.isLoading) {
    // acknowledgement during loading is invalid (should not happen)
    return state;
  }
  const next = { ...traceArchive, isAcknowledged: true };
  return { ...state, [traceID]: next };
}

function archiveStarted(state: TracesArchive, { meta }: ArchiveAction) {
  return { ...state, [meta.id]: { isLoading: true } };
}

function archiveDone(state: TracesArchive, { meta }: ArchiveAction) {
  return { ...state, [meta.id]: { isArchived: true, isAcknowledged: false } };
}

function archiveErred(state: TracesArchive, { meta, payload }: ArchiveAction) {
  if (!payload) {
    // make flow happy
    throw new Error('Invalid state, missing API error details');
  }
  const traceArchive = { error: payload, isArchived: false, isError: true, isAcknowledged: false };
  return { ...state, [meta.id]: traceArchive };
}

export default handleActions(
  {
    [actionTypes.ACKNOWLEDGE]: acknowledge,
    [`${archiveTrace}_PENDING`]: archiveStarted,
    [`${archiveTrace}_FULFILLED`]: archiveDone,
    [`${archiveTrace}_REJECTED`]: archiveErred,
  },
  initialState
);
