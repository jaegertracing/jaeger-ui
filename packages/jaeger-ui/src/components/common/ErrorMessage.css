/*
Copyright (c) 2017 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.ErrorMessage {
  position: relative;
  word-break: break-word;
  word-wrap: break-word;
}

.ErrorMessage--msg {
  color: #c00;
  margin: 0;
  padding-bottom: 1rem;
}

.ErrorMessage--details {
  background: #f5f5f5;
  border: 1px solid #e6e6e6;
  overflow: auto;
}

.ErrorMessage--detailsTable {
  min-width: 100%;
}

.ErrorMessage--detailItem:nth-child(2n) > .ErrorMessage--attr,
.ErrorMessage--detailItem:nth-child(2n) > .ErrorMessage--value {
  background-color: #f0f0f0;
  border-bottom: 1px solid #e8e8e8;
  border-top: 1px solid #e8e8e8;
}

/* dont have extra borders on the bottom row */
.ErrorMessage--detailItem:last-child > .ErrorMessage--attr,
.ErrorMessage--detailItem:last-child > .ErrorMessage--value {
  border-bottom: none;
}

.ErrorMessage--attr {
  color: rgba(0, 0, 0, 0.45);
  padding: 0.5em;
  vertical-align: top;
  white-space: nowrap;
}

.ErrorMessage--value {
  color: rgba(0, 0, 0, 0.8);
  font-family: monospace;
  padding: 0.5em;
  white-space: pre;
}
