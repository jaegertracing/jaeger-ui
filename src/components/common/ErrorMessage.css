/*
Copyright (c) 2017 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.ErrorMessage {
  background: #f5f5f5;
  border: 1px solid #e6e6e6;
  position: relative;
  word-break: break-word;
  word-wrap: break-word;
}

.ErrorMessage--msg {
  color: #c00;
  margin: 0;
  padding: 1rem;
}

.ErrorMessage--details {
  overflow: auto;
  padding: 0 1rem 1rem 1rem;
}

.ErrorMessage--attr {
  padding-bottom: 0.5em;
  padding-right: 1em;
  vertical-align: top;
  white-space: nowrap;
}

.ErrorMessage--value {
  font-family: monospace;
  white-space: pre;
}
