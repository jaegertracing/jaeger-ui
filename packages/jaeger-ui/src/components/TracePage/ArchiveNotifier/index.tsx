// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Icon, notification } from 'antd';

import ErrorMessage from '../../common/ErrorMessage';
import { TNil } from '../../../types';
import { TraceArchive } from '../../../types/archive';

import './index.css';

enum ENotifiedState {
  Progress = 'ENotifiedState.Progress',
  Outcome = 'ENotifiedState.Outcome',
}

// const NOTIFIED_PROGRESS = 'NOTIFIED_PROGRESS';
// const NOTIFIED_OUTCOME = 'NOTIFIED_OUTCOME';

// type NotifiedState = 'NOTIFIED_PROGRESS' | 'NOTIFIED_OUTCOME' | null;

type Props = {
  // eslint-disable-next-line react/no-unused-prop-types
  archivedState: TraceArchive | TNil;
  // eslint-disable-next-line react/no-unused-prop-types
  acknowledge: () => void;
};

type State = {
  notifiedState: ENotifiedState | null;
};

function getNextNotifiedState(props: Props) {
  const { archivedState } = props;
  if (!archivedState) {
    return null;
  }
  if (archivedState.isLoading) {
    return ENotifiedState.Progress;
  }
  return archivedState.isAcknowledged ? null : ENotifiedState.Outcome;
}

function updateNotification(oldState: ENotifiedState | null, nextState: ENotifiedState | null, props: Props) {
  if (oldState === nextState) {
    return;
  }
  if (oldState) {
    notification.close(oldState);
  }
  if (nextState === ENotifiedState.Progress) {
    notification.info({
      key: ENotifiedState.Progress,
      description: null,
      duration: 0,
      icon: <Icon type="loading" />,
      message: 'Archiving trace...',
    });
    return;
  }
  const { acknowledge, archivedState } = props;
  if (nextState === ENotifiedState.Outcome) {
    if (archivedState && archivedState.error) {
      const error = typeof archivedState.error === 'string' ? archivedState.error : archivedState.error;
      notification.warn({
        key: ENotifiedState.Outcome,
        className: 'ArchiveNotifier--errorNotification',
        message: <ErrorMessage.Message error={error} wrap />,
        description: <ErrorMessage.Details error={error} wrap />,
        duration: null,
        icon: <Icon type="clock-circle-o" className="ArchiveNotifier--errorIcon" />,
        onClose: acknowledge,
      });
    } else if (archivedState && archivedState.isArchived) {
      notification.success({
        key: ENotifiedState.Outcome,
        description: null,
        duration: null,
        icon: <Icon type="clock-circle-o" className="ArchiveNotifier--doneIcon" />,
        message: 'This trace has been archived.',
        onClose: acknowledge,
      });
    } else {
      throw new Error('Unexpected condition');
    }
  }
}

function processProps(notifiedState: ENotifiedState | null, props: Props) {
  const nxNotifiedState = getNextNotifiedState(props);
  updateNotification(notifiedState, nxNotifiedState, props);
  return nxNotifiedState;
}

export default class ArchiveNotifier extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    const notifiedState = processProps(null, props);
    this.state = { notifiedState };
  }

  static getDerivedStateFromProps(props: Props, state: State) {
    const notifiedState = processProps(state.notifiedState, props);
    return { notifiedState };
  }

  componentWillUnmount() {
    const { notifiedState } = this.state;
    if (notifiedState) {
      notification.close(notifiedState);
    }
  }

  render() {
    return null;
  }
}
