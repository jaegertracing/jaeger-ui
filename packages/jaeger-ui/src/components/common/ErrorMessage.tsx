// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';

import { ApiError } from '../../types/api-error';

import './ErrorMessage.css';

type ErrorMessageProps = {
  className?: string;
  detailClassName?: string;
  messageClassName?: string;
  error: ApiError;
};

type SubPartProps = {
  className?: string;
  error: ApiError;
  wrap?: boolean;
  wrapperClassName?: string;
};

type ErrorAttrProps = {
  name: string;
  value: string | number;
};

export const MAX_DETAIL_LENGTH = 1024;

function ErrorAttr({ name, value }: ErrorAttrProps) {
  return (
    <tr className="ErrorMessage--detailItem">
      <td className="ErrorMessage--attr">{name}</td>
      <td className="ErrorMessage--value">{value}</td>
    </tr>
  );
}

export function Message(props: SubPartProps) {
  const { className, error, wrap, wrapperClassName } = props;
  const cssClass = `ErrorMessage--msg ${className || ''}`;

  const msg =
    typeof error === 'string' ? (
      <h3 className={cssClass}>{error}</h3>
    ) : (
      <h3 className={cssClass}>{error.message}</h3>
    );

  if (wrap) {
    return <div className={`ErrorMessage ${wrapperClassName || ''}`}>{msg}</div>;
  }

  return msg;
}

Message.defaultProps = {
  className: undefined,
  wrap: false,
  wrapperClassName: undefined,
};

export function Details(props: SubPartProps) {
  const { className, error, wrap, wrapperClassName } = props;

  if (typeof error === 'string') {
    return null;
  }

  const { httpStatus, httpStatusText, httpUrl, httpQuery, httpBody } = error;
  const bodyExcerpt =
    httpBody && httpBody.length > MAX_DETAIL_LENGTH
      ? `${httpBody.slice(0, MAX_DETAIL_LENGTH - 3).trim()}...`
      : httpBody;

  const details = (
    <div
      className={`ErrorMessage--details ${className || ''} u-simple-scrollbars`}
      data-testid="ErrorMessage--details"
    >
      <table className="ErrorMessage--detailsTable">
        <tbody>
          {httpStatus && <ErrorAttr name="Status" value={httpStatus} />}
          {httpStatusText && <ErrorAttr name="Status text" value={httpStatusText} />}
          {httpUrl && <ErrorAttr name="URL" value={httpUrl} />}
          {httpQuery && <ErrorAttr name="Query" value={httpQuery} />}
          {bodyExcerpt && <ErrorAttr name="Response body" value={bodyExcerpt} />}
        </tbody>
      </table>
    </div>
  );

  if (wrap) {
    return (
      <div className={`ErrorMessage ${wrapperClassName || ''}`} data-testid="ErrorMessage--details--wrapper">
        {details}
      </div>
    );
  }

  return details;
}

Details.defaultProps = {
  className: undefined,
  wrap: false,
  wrapperClassName: undefined,
};

export default function ErrorMessage({
  className,
  detailClassName,
  error,
  messageClassName,
}: ErrorMessageProps) {
  if (!error) {
    return null;
  }

  if (typeof error === 'string') {
    return <Message className={messageClassName} error={error} wrapperClassName={className} wrap />;
  }

  return (
    <div className={`ErrorMessage ${className || ''}`} data-testid="ErrorMessage">
      <Message error={error} className={messageClassName} />
      <Details error={error} className={detailClassName} />
    </div>
  );
}

ErrorMessage.defaultProps = {
  className: undefined,
  detailClassName: undefined,
  messageClassName: undefined,
};
