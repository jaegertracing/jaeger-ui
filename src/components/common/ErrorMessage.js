// @flow

// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';

import './ErrorMessage.css';

type ErrorMessageProps = {
  className?: string,
  error:
    | string
    | {
        message: string,
        httpStatus?: any,
        httpStatusText?: string,
        httpUrl?: string,
        httpQuery?: string,
        httpBody?: string,
      },
};

function ErrorAttr({ name, value }: { name: string, value: any }) {
  return (
    <tr>
      <td className="ErrorMessage--attr">{name}</td>
      <td className="ErrorMessage--value">{value}</td>
    </tr>
  );
}

export default function ErrorMessage({ className, error }: ErrorMessageProps) {
  if (!error) {
    return null;
  }
  if (typeof error === 'string') {
    return (
      <div className={`ErrorMessage ${className || ''}`}>
        <h3 className="ErrorMessage--msg">{error}</h3>
      </div>
    );
  }
  const { message, httpStatus, httpStatusText, httpUrl, httpQuery, httpBody } = error;
  const bodyExcerpt = httpBody && httpBody.length > 1024 ? `${httpBody.slice(0, 1021).trim()}...` : httpBody;
  return (
    <div className={`ErrorMessage ${className || ''}`}>
      <h2 className="ErrorMessage--msg">{message}</h2>
      <div className="ErrorMessage--details u-simple-scrollbars">
        <table>
          <tbody>
            {httpStatus ? <ErrorAttr name="Status" value={httpStatus} /> : null}
            {httpStatusText ? <ErrorAttr name="Status text" value={httpStatusText} /> : null}
            {httpUrl ? <ErrorAttr name="URL" value={httpUrl} /> : null}
            {httpQuery ? <ErrorAttr name="Query" value={httpQuery} /> : null}
            {bodyExcerpt ? <ErrorAttr name="Response body" value={bodyExcerpt} /> : null}
          </tbody>
        </table>
      </div>
    </div>
  );
}

ErrorMessage.defaultProps = {
  className: undefined,
};
