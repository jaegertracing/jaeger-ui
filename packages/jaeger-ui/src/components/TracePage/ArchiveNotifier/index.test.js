// Copyright (c) 2020 The Jaeger Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { mount, shallow } from 'enzyme';
import { Icon, notification } from 'antd';
import ArchiveNotifier from './index';
import ErrorMessage from '../../common/ErrorMessage';

jest.mock('antd', () => {
  const originalModule = jest.requireActual('antd');

  return {
    __esModule: true, // Use it when dealing with esModules
    ...originalModule,
    notification: {
      close: jest.fn(),
      info: jest.fn(),
      success: jest.fn(),
      warn: jest.fn(),
    },
  };
});

describe('<ArchiveNotifier>', () => {
  let wrapper;
  let defaultProps;

  beforeEach(() => {
    defaultProps = {
      archivedState: { isArchived: true },
      acknowledge: jest.fn(),
    };

    wrapper = mount(<ArchiveNotifier {...defaultProps} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders with default props', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('notification success() is called with default props', () => {
    expect(notification.success).toBeCalledWith(
      expect.objectContaining({
        key: 'ENotifiedState.Outcome',
        description: null,
        duration: null,
        icon: <Icon className="ArchiveNotifier--doneIcon" type="clock-circle-o" />,
        message: 'This trace has been archived.',
        onClose: defaultProps.acknowledge,
      })
    );
  });

  it('notification close() is called onUnmount', () => {
    wrapper.unmount();
    expect(notification.close).toBeCalledWith('ENotifiedState.Outcome');
  });

  it('notification close() is not called onUnmount of null state', () => {
    const props = { ...defaultProps, archivedState: null };
    wrapper = mount(<ArchiveNotifier {...props} />);

    expect(wrapper.state().notifiedState).toEqual(null);
    wrapper.unmount();
    expect(notification.close).not.toBeCalled();
  });

  it('notification close() is not called onUnmount of isAcknowledged state', () => {
    const props = { ...defaultProps, archivedState: { isAcknowledged: true } };
    wrapper = mount(<ArchiveNotifier {...props} />);

    wrapper.unmount();
    expect(notification.close).not.toBeCalled();
  });

  it('will throw on missing both isArchived and error from archivedState prop', () => {
    const props = { ...defaultProps, archivedState: {} };
    expect(() => {
      shallow(<ArchiveNotifier {...props} />);
    }).toThrow('Unexpected condition');
  });

  /**
   * TODO: Is there a better way to test these next two?
   *
   * state is internal to the component and probably shouldn't be checked externally, but in this
   * case, this is basically a no-op, and the test would be a potentially infinite list of things
   * that didn't occur, which isn't a great test either.
   */
  it('sets internal notifiedState to null on null archivedState prop', () => {
    const props = { ...defaultProps, archivedState: null };
    wrapper = mount(<ArchiveNotifier {...props} />);

    expect(wrapper.state().notifiedState).toEqual(null);
  });

  it('sets internal notifiedState to null on archivedState.isAcknowledged', () => {
    const props = { ...defaultProps, archivedState: { isAcknowledged: true } };
    wrapper = mount(<ArchiveNotifier {...props} />);

    expect(wrapper.state().notifiedState).toEqual(null);
  });

  it('will call notification.info on isLoading is true', () => {
    const props = { ...defaultProps, archivedState: { isLoading: true } };
    wrapper = mount(<ArchiveNotifier {...props} />);

    expect(notification.info).toBeCalledWith(
      expect.objectContaining({
        key: 'ENotifiedState.Progress',
        description: null,
        duration: 0,
        icon: <Icon type="loading" />,
        message: 'Archiving trace...',
      })
    );
  });

  it('will call notification.warn on error state', () => {
    const props = { ...defaultProps, archivedState: { error: 'This is an error string' } };
    wrapper = mount(<ArchiveNotifier {...props} />);

    expect(notification.warn).toBeCalledWith(
      expect.objectContaining({
        key: 'ENotifiedState.Outcome',
        className: 'ArchiveNotifier--errorNotification',
        description: <ErrorMessage.Details error="This is an error string" wrap />,
        duration: null,
        icon: <Icon className="ArchiveNotifier--errorIcon" type="clock-circle-o" />,
        message: <ErrorMessage.Message error="This is an error string" wrap />,
        onClose: props.acknowledge,
      })
    );
  });

  it('will call notification.close new state type', () => {
    const props = { ...defaultProps, archivedState: { isLoading: true } };
    wrapper = mount(<ArchiveNotifier {...props} />);

    const newProps = { ...props, archivedState: { isArchived: true } };
    wrapper.setProps(newProps);

    expect(notification.close).toBeCalledWith('ENotifiedState.Progress');
  });

  it('will call notification.close on null state update', () => {
    const newProps = { ...defaultProps, archivedState: null };
    wrapper.setProps(newProps);

    expect(notification.close).toBeCalledWith('ENotifiedState.Outcome');
  });
});
