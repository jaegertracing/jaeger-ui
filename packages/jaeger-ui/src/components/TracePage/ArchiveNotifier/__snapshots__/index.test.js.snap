// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArchiveNotifier> renders with default props 1`] = `""`;
